### ISSUE TITLE
Openstack cinder storageclass should support also random zone names.

### ISSUE THREAD

**Issue by zetaab** [kubernetes/kubernetes#44735]
**Is this a BUG REPORT or FEATURE REQUEST?** (choose one): FEATURE REQUEST

**Kubernetes version** (use `kubectl version`): 1.6.1

**Environment**:
- **Cloud provider or hardware configuration**: openstack
- **OS** (e.g. from /etc/os-release): centos7

**What happened**:

We have openstack environment with three different zones, called as `zone-1`, `zone-2` and `zone-3`. The problem with current storageclass is that it does not support any random method to provision storages to different zones. Like mentioned here https://github.com/kubernetes/kubernetes/issues/35695#issuecomment-277165783 other classes have some support to that. In openstack if we do not define availability zone, it will always use zone called `nova`.

Example: We want provision mysql galera 3 node cluster, that we have one node in each zone. That is impossible with kubernetes currently because we need to always defined in which zone our storageclass is pointing.

```
kind: StorageClass
apiVersion: storage.k8s.io/v1beta1
metadata:
  name: zone1
provisioner: kubernetes.io/cinder
parameters:
  availability: zone-1
---
apiVersion: v1
kind: Service
metadata:
  annotations:
    service.alpha.kubernetes.io/tolerate-unready-endpoints: "true"
  name: galera
  labels:
    app: mysql
spec:
  ports:
  - port: 3306
    name: mysql
  clusterIP: None
  selector:
    app: mysql
---
apiVersion: apps/v1beta1
kind: StatefulSet
metadata:
  name: mysql
spec:
  serviceName: "galera"
  replicas: 3
  template:
    metadata:
      labels:
        app: mysql
      annotations:
        pod.alpha.kubernetes.io/initialized: "true"
    spec:
      containers:
      - name: mysql
        image: adfinissygroup/k8s-mariadb-galera-centos:v002
        imagePullPolicy: Always
        ports:
        - containerPort: 3306
          name: mysql
        - containerPort: 4444
          name: sst
        - containerPort: 4567
          name: replication
        - containerPort: 4568
          name: ist
        readinessProbe:
          exec:
            command:
            - /usr/share/container-scripts/mysql/readiness-probe.sh
          initialDelaySeconds: 15
          timeoutSeconds: 5
        volumeMounts:
        - name: datadir
          mountPath: /var/lib/mysql
        env:
          - name: POD_NAMESPACE
            valueFrom:
              fieldRef:
                apiVersion: v1
                fieldPath: metadata.namespace
  volumeClaimTemplates:
  - metadata:
      name: datadir
      annotations:
       volume.beta.kubernetes.io/storage-class: zone1
    spec:
      accessModes: [ "ReadWriteOnce" ]
      resources:
        requests:
          storage: 2Gi
```

There are actually two problems in this yaml: 1) storageclass zone name must be hardcoded, we do not have way to spread statefulset to multiple AZs. 2) openstack plugin do not have skill to read in which zone PVC is located. If I will execute this yaml, all pvcs are created to zone-1 but it might be that some stateful pods are going to zone-2 or to zone-3.

**What you expected to happen**: I except that we should have somekind of way to create multi AZ statefulsets. BTW do we have way to say "please spread statefulset replicas to all zones" kind of similar feature than daemonset in nodes, but for zones.

**Comment by zetaab** [comment]
sig/openstack sig/storage and maybe area/stateful-apps labels needed

**Comment by zetaab** [comment]
@justinsb could you help little bit here. I can easily modify that if storageclass availability zone is not defined, it will use similar technique like here https://github.com/kubernetes/kubernetes/blob/2cb17cc67745aa39f700e1d0c3d22c70074bee46/pkg/volume/util.go#L287 to choose random zone. However, in my tests it does not solve that second problem. If I create storageclass cinder to zone-1 it might be that statefulset is creating pods to zone-2&zone-3 as well. How we could get support for that? I mean that pod should check "ok the pvc is created zone-x i should also provision pod to similar zone"

**Comment by saad-ali** [comment]
CC @kubernetes/sig-storage-feature-requests @kubernetes/sig-openstack-feature-requests

**Comment by justinsb** [comment]
@zetaab looks like you solved it in #44798 - sorry I didn't get to this sooner :-)

**Comment by jianglingxia** [comment]
How does k8s use this kind of storage of cinder to store ? How do I configure this environment?can you help me ?thanks very much @zetaab 
only use the file ?:
# cat /etc/cloud.conf
[Global]
username=<user>
password=XXXXXXXX
auth-url=http://<openStack URL>:5000/v2.0
tenant-name=Shadow
region=RegionOne
